TRACING.md

```diff
@@ -0,0 +1,312 @@
+
+- Container manager requested
+
+  The container manager requests the container be stopped, for example, by
+  the user typing the following:
+
+  ```
+  $ sudo docker stop $container_id
+  ```
+
+  This results in the container manager calling `kata-runtime kill $container_id`,
+  which causes the following to happen:
+
+  - The `kata-agent` `DestroySandbox` gRPC API is called.
+  - The runtime sends a [QMP](https://wiki.qemu.org/Documentation/QMP) `quit`
+    command to the VM to shut it down.
+
+# Adding tracing support to the agent
+
+The crucial difference between the two shutdown methods explained in the [VM
+shutdown](#vm-shutdown) section is that the user-initiated shutdown does not
+explicitly destroy the sandbox using the `DestroySandbox` gRPC API: the
+sandbox is "destroyed" by the agent shutting down the VM.
+
+Since tracing should work in all scenarios, it cannot be triggered in the
+obvious location of `DestroySandbox`. Therefore, the approach taken is for the
+agent to trigger the end of tracing when the primary workload process exits by
+making `WaitProcess` trigger the gRPC server to end.
+
+# Guest to Host communication using VSOCK
+
+Since [Jaeger does not yet support VSOCK](https://github.com/jaegertracing/jaeger/issues/1019),
+it is necessary to provide a way for the Jaeger client code (running inside the `kata-agent`
+in the VM), to connect to the Jaeger agent (running on the host). The default
+Jaeger port for such communications is
+[UDP port `6831`](https://www.jaegertracing.io/docs/getting-started/).
+
+The required redirection is a achieved using two instances of the `socat(1)` network utility:
+
+- The first instance runs in the guest environment in the VM and forwards all
+  Jaeger trace traffic out of the VM using a VSOCK socket.
+
+- The second instance runs in the host environment and forwards all Jaeger
+  traffic from the VSOCK socket to the Jaeger agent's UDP port.
+
+Further details:
+
+| Name | Type | Environment | Summary | Details |
+|-|-|-|-|-|
+| `jaeger-client-socat-redirector.service` | systemd service | Guest | Guest `UDP:6831` to guest `VSOCK:6831` | Redirects data sent to UDP port `6831` to the host context `VSOCK` socket on port `6831`. |
+| `vsock-to-udp-server.sh` | shell script | Host | Host `VSOCK:6831` to host `UDP:6831` | Redirects all host context VSOCK data on port `6831` to the Jaeger agent, which is listening on UDP port `6831`. |
+
+> **Note:**
+>
+> This is a temporary solution until Jaeger supports VSOCK natively.
+
+For tracing to work, the host system must have a Jaeger agent running.
+
+# Enabling tracing
+
+Tracing is enabled by adding the `agent.trace` option in the `kernel_params=`
+variable in the runtime's `configuration.toml` file.
+
+## Tracing options
+
+Tracing support in the agent is either disabled, or set to one of the
+following:
+
+| Trace type | Trace setting | Description | Use-case | Notes |
+|-|-|-|-|
```

Comment:
This table is not rendering for me - I suspect you are missing one set of `| - |` in the second line.

Reply:
Weirdly it rendered for me in both github and atom. What are you using ooi?

Reply:
I was just looking in the 'view file' on the github PR - so, dunno why that was different from your github view...